test/integration/job/job_test.go

```diff
@@ -1820,6 +1820,137 @@ func TestJobPodReplacementPolicy(t *testing.T) {
+		},
+		"podReplacementPolicy terminatingOrFailed; restart and disable feature; renable feature": {
+			podReplacementPolicyEnabled: true,
+			jobSpec: &batchv1.JobSpec{
+				Parallelism:          pointer.Int32Ptr(podCount),
+				Completions:          pointer.Int32Ptr(podCount),
+				CompletionMode:       &nonIndexedCompletion,
+				PodReplacementPolicy: podReplacementPolicy(batchv1.TerminatingOrFailed),
+			},
+			wantTerminating: pointer.Int32(podCount),
+			wantActive:      int(podCount),
+			wantFailed:      int(podCount),
+		},
+	}
+	for name, tc := range cases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodReplacementPolicy, true)()
+
+			closeFn, restConfig, clientSet, ns := setup(t, "pod-replacement-policy")
+			defer closeFn()
+			ctx, cancel := startJobControllerAndWaitForCaches(restConfig)
+			defer func() {
+				cancel()
+			}()
+			resetMetrics()
+
+			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &batchv1.Job{
+				Spec: *tc.jobSpec,
+			})
+			if err != nil {
+				t.Fatalf("Failed to create Job: %v", err)
+			}
+			jobClient := clientSet.BatchV1().Jobs(jobObj.Namespace)
+
+			// Wait for pods to start up.
+			err = wait.PollImmediate(5*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+				job, err := jobClient.Get(ctx, jobObj.Name, metav1.GetOptions{})
+				if err != nil {
+					return false, err
+				}
+				if job.Status.Active == int32(podCount) {
+					return true, nil
+				}
+				return false, nil
+			})
+			if err != nil {
+				t.Fatalf("Error waiting for Job pods to become active: %v", err)
+			}
+			pods, errList := clientSet.CoreV1().Pods(ns.Namespace).List(ctx, metav1.ListOptions{})
+			if errList != nil {
+				t.Fatalf("Failed to list pods: %v", errList)
+			}
+			updatePod(t, clientSet, pods.Items, func(pod *v1.Pod) {
+				pod.Finalizers = append(pod.Finalizers, "fake.example.com/blockDeletion")
+			})
+			err = clientSet.CoreV1().Pods(ns.Name).DeleteCollection(ctx,
+				metav1.DeleteOptions{},
+				metav1.ListOptions{
+					Limit: 1000,
+				})
+			if err != nil {
+				t.Fatalf("Failed to cleanup Pods: %v", err)
+			}
+
+			podsDelete, errList2 := clientSet.CoreV1().Pods(ns.Namespace).List(ctx, metav1.ListOptions{})
+			if errList != nil {
+				t.Fatalf("Failed to list pods: %v", errList2)
+			}
+			for _, val := range podsDelete.Items {
+				if val.DeletionTimestamp == nil {
+					t.Fatalf("Deletion not registered.")
+				}
+			}
+
+			validateJobsPodsStatusOnly(ctx, t, clientSet, jobObj, podsByStatus{
+				Terminating: tc.wantTerminating,
+				Failed:      tc.wantFailed,
+				Ready:       pointer.Int32(0),
+			})
+			// Disable the controller and turn off feature.
+			cancel()
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodReplacementPolicy, false)()
+			ctx, cancelDisabled := startJobControllerAndWaitForCaches(restConfig)
```

Comment:
I would suggest to name it `cancel`, since we reuse the name `ctx`. This is what we do in other tests in this file, like here: https://github.com/kubernetes/kubernetes/blob/887ac275a2cfb6030df80bbe3066609336fa220a/test/integration/job/job_test.go#L448